Regression: indexing by `np.int64` indices

The Array API [states](https://data-apis.org/array-api/latest/API_specification/indexing.html):

> An integer index must be an object satisfying [operator.index](https://www.python.org/dev/peps/pep-0357/) (e.g., int).

Therefore, this is legal, because `operator.index(np.int64(0))` works:
```python
>>> import array_api_strict as xp
>>> import numpy as np
>>> a = xp.asarray([10, 20])
>>> a[np.int64(0)]
10
```

#139 broke it and caused a regression in scipy:
```
FAILED scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans2_init[array_api_strict] - ValueError: Array indexing is only allowed when array to be indexed and all indexing arrays are on the same device.
FAILED scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans2_kpp_low_dim[array_api_strict] - ValueError: Array indexing is only allowed when array to be indexed and all indexing arrays are on the same device.
FAILED scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans2_kpp_high_dim[array_api_strict] - ValueError: Array indexing is only allowed when array to be indexed and all indexing arrays are on the same device.
FAILED scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans_and_kmeans2_random_seed[array_api_strict] - ValueError: Array indexing is only allowed when array to be indexed and all indexing arrays are on the same device.
```

CC @lucyleeow

--- a/array_api_strict/_array_object.py
+++ b/array_api_strict/_array_object.py
@@ -722,7 +722,7 @@ class Array:
         devices = {self.device}
         if isinstance(key, tuple):
             devices.update(
-                [subkey.device for subkey in key if hasattr(subkey, "device")]
+                [subkey.device for subkey in key if isinstance(subkey, Array)]
             )
             if len(devices) > 1:
                 raise ValueError(

--- a/array_api_strict/tests/test_array_object.py
+++ b/array_api_strict/tests/test_array_object.py
@@ -106,7 +106,6 @@ def test_validate_index():
     "integer_index",
     [
         1,
-        np.bool(1),
         np.int8(0),
         np.uint8(0),
         np.int16(0),

--- a/array_api_strict/tests/test_array_object.py
+++ b/array_api_strict/tests/test_array_object.py
@@ -102,6 +102,31 @@ def test_validate_index():
 
 
 @pytest.mark.parametrize("device", [None, "CPU_DEVICE", "device1", "device2"])
+@pytest.mark.parametrize(
+    "integer_index",
+    [
+        1,
+        np.bool(1),
+        np.int8(0),
+        np.uint8(0),
+        np.int16(0),
+        np.uint16(0),
+        np.int32(0),
+        np.uint32(0),
+        np.int64(0),
+        np.uint64(0),
+        2,
+    ],
+)
+def test_indexing_ints(integer_index, device):
+    # Ensure indexing with different integer types works on all Devices.
+    device = None if device is None else Device(device)
+
+    a = arange(5, device=device)
+    a[integer_index]
+
+
+@pytest.mark.parametrize("device", [None, "CPU_DEVICE", "device1", "device2"])
 def test_indexing_arrays(device):
     # indexing with 1D integer arrays and mixes of integers and 1D integer are allowed
     device = None if device is None else Device(device)

--- a/array_api_strict/tests/test_array_object.py
+++ b/array_api_strict/tests/test_array_object.py
@@ -100,12 +100,18 @@ def test_validate_index():
     assert_raises(IndexError, lambda: a[:])
     assert_raises(IndexError, lambda: a[idx])
 
+class DummyIndex:
+    def __init__(self, x):
+        self.x = x
+    def __index__(self):
+        return self.x
+
 
 @pytest.mark.parametrize("device", [None, "CPU_DEVICE", "device1", "device2"])
 @pytest.mark.parametrize(
     "integer_index",
     [
-        1,
+        0,
         np.int8(0),
         np.uint8(0),
         np.int16(0),
@@ -114,6 +120,7 @@ def test_validate_index():
         np.uint32(0),
         np.int64(0),
         np.uint64(0),
+        DummyIndex(0),
     ],
 )
 def test_indexing_ints(integer_index, device):
@@ -121,7 +128,7 @@ def test_indexing_ints(integer_index, device):
     device = None if device is None else Device(device)
 
     a = arange(5, device=device)
-    a[integer_index]
+    assert a[(integer_index,)] == a[integer_index] == a[0]
 
 
 @pytest.mark.parametrize("device", [None, "CPU_DEVICE", "device1", "device2"])

--- a/array_api_strict/tests/test_array_object.py
+++ b/array_api_strict/tests/test_array_object.py
@@ -115,7 +115,6 @@ def test_validate_index():
         np.uint32(0),
         np.int64(0),
         np.uint64(0),
-        2,
     ],
 )
 def test_indexing_ints(integer_index, device):

